Prune with glorot uniform normalization

dorimedini opened the pull request:
Addresses #104 , rebased on top of #102

Got new results, still can't see correlation though... winnery intersection and robustness don't follow the same pattern

*EDIT:* fixed some bugs, got results that make sense. Still no correlation:

![image](https://user-images.githubusercontent.com/3027929/63580216-ffe0bf80-c59c-11e9-9453-163991fc99f9.png)
![image](https://user-images.githubusercontent.com/3027929/63580222-040cdd00-c59d-11e9-80d4-1cb89ff14559.png)
![image](https://user-images.githubusercontent.com/3027929/63580231-096a2780-c59d-11e9-917d-976334f4a9ad.png)
![image](https://user-images.githubusercontent.com/3027929/63580241-0bcc8180-c59d-11e9-9733-5b7f72d67936.png)

dorimedini committed:
Allow pruning without pruning in-place

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Can now compute the glorot constant

See [glorot_uniform](https://keras.io/initializers/#glorot_uniform) in the keras docs

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Pruning now biases the prune threshold by glorot constant

This is default behavior, can be suppressed by flag (override _prune in
inheriting classes)

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
New prune sweetspot

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Fix glorot constant computation

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Trainers compute list of glorot-drawn layers

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
When pruning you must now provide glorot layer list

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
New magic constant (just 50% untraied prune)

Signed-off-by: dorimedini <dorimedini@gmail.com>

galshachaf commented on trainer.py:
```diff
@@ -260,6 +273,18 @@ def get_weighted_layers_indices(self):
+        if not (layer_index in self.get_weighted_layers_indices()):
+            self.logger.debug("Layer {} not weighted".format(layer_index))
+            return False
+        ret = (not self._regularizer) or self._regularizer == 'glorot_uniform'
```
regularizer or initializer?

dorimedini replied:
Fixed (commit 84da80a)

dorimedini committed:
Fixed incorrect implementation of is_glorot

regularizer != initializer. FCTrainer doesn't provide API for custom
kernel initializer, so just return true for any Dense layer (any
weighted layer in this case).
Also removes some debug prints

Signed-off-by: dorimedini <dorimedini@gmail.com>
